From 04b63bdff620b4704548605ec8545e3a704810ad Mon Sep 17 00:00:00 2001
From: Binh Nguyen <binh.nguyen.uw@renesas.com>
Date: Fri, 13 Sep 2019 15:47:41 +0700
Subject: [PATCH] plat: renesas: rcar: bl2_fusa: Add option to enable ECC for
 full DRAM

This commit add an option to enable ECC for full DRAM on EK874 with
mode 8bits data/5bits ECC.

To enable ECC for full DRAM set RZG_DRAM_ECC_FULL=1 in build options.
Below is memory map when RZG_DRAM_ECC_FULL=1 :

 ---------
| Security| 128 MB
 ---------
| Data    | 960 MB
 ---------
| ECC code| 960 MB
 ---------

Signed-off-by: Binh Nguyen <binh.nguyen.uw@renesas.com>
---
 plat/renesas/rcar/bl2_fusa.c  | 13 +++++++++++++
 plat/renesas/rcar/platform.mk |  6 ++++++
 2 files changed, 19 insertions(+)

diff --git a/plat/renesas/rcar/bl2_fusa.c b/plat/renesas/rcar/bl2_fusa.c
index 290565a..0bc2e37 100644
--- a/plat/renesas/rcar/bl2_fusa.c
+++ b/plat/renesas/rcar/bl2_fusa.c
@@ -28,11 +28,19 @@ static const struct rzg2_ecc_conf {
 	uint32_t fusaareacr;
 	uint32_t eccareacr;
 } rzg2_ek874_conf[] = {
+#if (RZG_DRAM_ECC_FULL == 1)
+	{ FUSAAREACR(1, 7, 0x408000000),    ECCAREACR(0, 0x448000000)   }, /* 128+128 MB */
+	{ FUSAAREACR(1, 8, 0x410000000),    ECCAREACR(0, 0x450000000)   }, /* 256+256 MB */
+	{ FUSAAREACR(1, 8, 0x420000000),    ECCAREACR(0, 0x460000000)   }, /* 256+256 MB */
+	{ FUSAAREACR(1, 8, 0x430000000),    ECCAREACR(0, 0x470000000)   }, /* 256+256 MB */
+	{ FUSAAREACR(1, 6, 0x440000000),    ECCAREACR(0, 0x444000000)   }, /* 64+64 MB */
+#else //(RZG_DRAM_ECC_FULL == 1)
 	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
 	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
 	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
 	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
 	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
+#endif //(RZG_DRAM_ECC_FULL == 1)
 	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
 	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
 	{ FUSAAREACR(0, 0, 0),    ECCAREACR(0, 0)   },
@@ -59,6 +67,11 @@ void bl2_enable_ecc_ek874(void)
 	for (n = 0; n < nb_of_conf; n++, conf++) {
 		mmio_write_32((uintptr_t)((uint32_t *)DFUSAAREACR + n), conf->fusaareacr);
 		mmio_write_32((uintptr_t)((uint32_t *)DECCAREACR + n), conf->eccareacr);
+		if(conf->fusaareacr & (1 << 31))
+			NOTICE("BL2: DRAM ECC area=%d, FuSa=0x%x ECC=0x%x, size=%d MB\n",
+					n, conf->fusaareacr, conf->eccareacr,
+					1 << ((conf->fusaareacr >> 24) & 0xf));
+
 	}
 
 }
diff --git a/plat/renesas/rcar/platform.mk b/plat/renesas/rcar/platform.mk
index 25d7e7e..94a0ae2 100644
--- a/plat/renesas/rcar/platform.mk
+++ b/plat/renesas/rcar/platform.mk
@@ -373,6 +373,12 @@ RZG_DRAM_EK874_ECC :=0
 endif
 $(eval $(call add_define,RZG_DRAM_EK874_ECC))
 
+# Process RZG_DRAM_ECC_FULL flag
+ifndef RZG_DRAM_ECC_FULL
+RZG_DRAM_ECC_FULL :=0
+endif
+$(eval $(call add_define,RZG_DRAM_ECC_FULL))
+
 # Process RCAR_BL33_ARG0 flag
 ifdef RCAR_BL33_ARG0
 $(eval $(call add_define,RCAR_BL33_ARG0))
-- 
2.7.4

